Make FormulaEngine instances composable, instead of FormulaReceivers

sahas-subramanian-frequenz opened the pull request:
This PR changes how formulas are composed -
  - receivers from formulas are no longer composable.
  - formula composition is now done by composing FormulaEngine instances.
  - Automatic formulas from the logical meter and *pools, are now properties, and return `FormulaEngine` instances, which can be composed further, or can provide a receiver to fetch values.

  ``` python
  grid_power_receiver = microgrid.logical_meter().grid_power.new_receiver()

  self._inverter_power = (
      microgrid.logical_meter().pv_power
      + microgrid.logical_meter().battery_power
  ).build("inverter_power")

  inverter_power_receiver = self._inverter_power.new_receiver()
  ```

The type hint issues in the formula engine are also resolved.

Based on #336.

shsms committed:
Drop mypy exclusion rule for formula engine

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

mathias-baumann-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine.py:
```diff
@@ -436,21 +437,21 @@ def __init__(
         return self._engine
 
-    def new_receiver(
+    def new_receiver(  # type: ignore[override]
```
Why is this necessary?

sahas-subramanian-frequenz replied:
Because `_BaseFormulaChannel` inherits `Broadcast`,  we should inherit the type of its methods as well,  i.e. this `new_receiver` should return `Receiver[Sample]`, etc.  But we are returning `FormulaReceiver` which derives from `Receiver[Sample]`,  but is still a different type with additional methods.  and mypy doesn't like this.

sahas-subramanian-frequenz replied:
This is no longer relevant,  because `FormulaReceiver` are gone now.  FormulaEngines directly return broadcast receivers.

sahas-subramanian-frequenz commented:
Trying to get https://github.com/frequenz-floss/frequenz-sdk-python/issues/314 into this PR as well,  draft until then.

shsms committed:
Remove custom formula channels and receivers

We will use regular broadcast channels, and provide composability on
FormulaEngine instances instead.

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Add a ComposableFormulaEngine base class

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Make FormulaEngine and FormulaEngine3Phase composable

- Also make `FormulaEngine` take a `FormulaBuilder` as constructor
  param, so that subscriptions are made directly from the
  FormulaEngine's task,  thereby avoiding async calls when creating
  formulas.

- Also remove all generic types, and use Union types instead.
  Overloads will be added in a subsequent commit, to fix any remaining
  typing issues as well.

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Improve typing for HigherOrderFormulaBuilder by adding overloads

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Update formula generators to work with new FormulaEngine interface

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Update remaining code to work with FormulaEngine interface changes

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Update tests to work with new formula engine interface

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Update RELEASE_NOTES.md

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

shsms committed:
Convert generated formula methods into properties

Signed-off-by: Sahas Subramanian <sahas.subramanian@frequenz.com>

sahas-subramanian-frequenz commented:
This is ready for review again.

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine.py:
```diff
@@ -473,76 +495,85 @@ def push_average(self, metrics: List[Tuple[str, Receiver[Sample], bool]]) -> Non
-        self._steps.append((TokenType.COMPONENT_METRIC, recv.clone()))
-        recv._deactivate()  # pylint: disable=protected-access
-        self._engine = None
+        self._engine: FormulaEngine | FormulaEngine3Phase
```
is it really needed to keep the state for this attribute?

sahas-subramanian-frequenz replied:
true,  it is not necessary.  removed.

daniel-zullo-frequenz reviewed (commented):
I only have a comment to check for

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py:
```diff
@@ -73,7 +68,7 @@ async def from_string(
         channel_key = formula + component_metric_id.value
         if channel_key in self._engines:
-            return self._engines[channel_key].new_receiver()
+            return self._engines[channel_key]  # type: ignore
```
Why does this requires a type ignore?

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_ev_charger_pool.py:
```diff
@@ -104,40 +104,47 @@ def component_ids(self) -> abc.Set[int]:
             EVChargerCurrentFormula,
             FormulaGeneratorConfig(component_ids=self._component_ids),
-        )
+        )  # type: ignore[return-value]
```
Also here or world be nice to clarify went the ignore is needed.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_ev_charger_pool.py:
```diff
@@ -104,40 +104,47 @@ def component_ids(self) -> abc.Set[int]:
             EVChargerPowerFormula,
             FormulaGeneratorConfig(component_ids=self._component_ids),
-        )
+        )  # type: ignore[return-value]
```
And here, etc.

leandro-lucarella-frequenz reviewed (commented):
Amazing to see so many async, await and new_receiver() go! 👋
